### DISCUSSION TITLE
Increase uplink slots

### DISCUSSION THREAD

**Discussion by jasonfarmer0992** [srsran/srsRAN_Project#252] [General Help] [answered]
Does anybody know if it is possible to increase the number of slots that are used for uplink and decrease the number of slots used for downlink in a 5G TDD setup? Can this be set in the configuration file, or is a code change required? It appears that the default slot allocation is 80% downlink and 20% uplink. I would like to flip that to favor uplink.

**Comment by brendan-mcauliffe** [comment]
Hi @jasonfarmer0992 if you take a look at the Configuration Reference page of the docs, [here](https://docs.srsran.com/projects/project/en/latest/user_manuals/source/config_ref.html), you will see there are parameters related to the TDD slot configuration under the sub-section `tdd_ul_dl_cfg` within the `cell-cfg` section.

**Comment by frankist** [comment]
As @brendan-mcauliffe said, you can change the TDD pattern in the config file.

We never tested a setup with more UL slots than DL slots, so let me know if you run into issues.

**Reply by HaoxinSEU** [reply]
Hi Francisco,

We tried to change this TDD pattern as you said. 

We changed `nof_dl_slots` to 3, and `nof_ul_slots` to 6, so we were allocating more slots to UL and less slots to DL, compared to the default setup. We expected that UL throughput would increase, while DL throughput would decrease. However, what we observed is that: DL throughput decreased as expected, but UL throughput didn't increase.

We also used different periods `dl_ul_tx_period` to do this test, but the results are always the same: the throughput of UL doesn't increase even if we allocate more slots.

Do you have any idea why this happens?

Thanks for your help!

Best regards

**Reply by ismagom** [reply]
If you send us logs, maybe we can help. Thanks

**Reply by HaoxinSEU** [reply]
Thanks for your reply! Could you be more specific, like which logs do you want? e.g. the configuration file, pcap, etc?

Thanks again!

Regards

**Reply by ismagom** [reply]
configuration file and gnb log with `mac_level=debug` and `phy_level=info` will be enough. 

thanks

**Reply by HaoxinSEU** [reply]
Hi,

We did more tests on this topic. And we found that in the past UL throughput didn't change because we had a bad `rx_gain`, so RX throughput is limited by this, no matter how many slots were allocated.

Then after using a correct `rx_gain`, we did two tests  with 20MHz BW to see the effects of setting different `nof_ul_slots`. It shows that UL throughput increases when increasing this value, but not in proportional to the number of slots allocated.

In the first test, we set `nof_ul_slots=3, nof_dl_slots=6`, and the UL and DL throughput are 14.2 and 56.5 Mbps. In the second test, we set `nof_ul_slots=6, nof_dl_slots=3`, and the UL and DL throughput are 20.7 and 28.1 Mbps. In both tests, the mixed slot is equally allocated to them (6 symbols for each). So you see that DL decreases by half as expected, while UL doesn't increase to x1.85 (in theory it should be 26 Mbps, right?).

Besides, when we change to 40MHz BW, both UL and DL throughput are in proportional to the number of slots allocated, as expected.

Below are the configurations and logs for the two tests with 20MHz BW. Thanks for your help!

[gnb_35PercentDownLink_65PercentUpLink.zip](https://github.com/srsran/srsRAN_Project/files/12893096/gnb_35PercentDownLink_65PercentUpLink.zip)
[gnb_65PercentDownLink_35PercentUpLink.zip](https://github.com/srsran/srsRAN_Project/files/12893097/gnb_65PercentDownLink_35PercentUpLink.zip)

**Reply by herlesupreeth** [reply]
@jasonfarmer0992 here is my two cents for increasing the number of uplink slots.

Just increasing the nof. UL slots will not do any good since DL slots are needed (sending PDCCH for UL DATA) to schedule UL data.

For example: Imagine 1 UE testing scenario and nof. DL slots is 3 and UL slots is 6. In this case the UE can be scheduled to transmit UL data in only 3 slots since the PDCCH to schedule UL data can be sent only over 3 slots so the remaining 3 slots are redundant.

So in my opinion there should be equal number of DL and UL slots.

**Reply by jasonfarmer0992** [reply]
@herlesupreeth I really appreciate your input, but I am pretty sure that you are incorrect. 

First, one of the biggest advantages of TDD is its ability to dynamically adjust the channel to favor uplink or downlink depending on the situation. If what you say is true, then there would be no reason to deal with the extra RF frontend and timing complexities of TDD, and we should always use FDD.

Second, you don't need one DL slot to schedule one UL slot. You do probably need at least one DL slot to schedule the UL slots. Here are some links to help understand how frames, slots, and symbols work in 5G NR:

- https://www.sharetechnote.com/html/5G/5G_FrameStructure.html
- https://www.sharetechnote.com/html/5G/5G_tdd_UL_DL_configurationCommon.html

I would highly recommend the ShareTechnote website on 5G NR. I am still trying to understand it all myself. It's pretty complicated, at least to me.

**Reply by herlesupreeth** [reply]
Hey @jasonfarmer0992 

> Second, you don't need one DL slot to schedule one UL slot. You do probably need at least one DL slot to schedule the UL slots.

In any given DL slot there can be only **one DCI per DCI format** for a DCI scrambled with **C-RNTI**. So there can be only one DCI (Format 0_1) in a DL slot scheduling the UL data. I would recommend reading TS 38.213 section 10.
NOTE: There can exist another DCI with DCI format 1_1 (alongside DCI format 0_1) in the same DL slot scheduling DL data

**Reply by frankist** [reply]
Hey @herlesupreeth, could that limit be circumvented with more than one CORESET per slot?

**Reply by herlesupreeth** [reply]
@frankist rather than CORESET, it could be alleviated by defining another SearchSpace with DCI format 1_0/0_0. This way we can schedule two UL slots one with DCI format 0_1 and another with DCI format 0_0 for the same C-RNTI.

**Reply by jasonfarmer0992** [reply]
@herlesupreeth thanks for the info! You are definitely smarter than me here.

Would it be possible to use the TDD-UL-DL-ConfigCommon to set a static configuration? This is discussed in section 11 of TS 38.213. I think it uses the RRC instead of PDCCH DCI. Can srsRAN already do this?

Also, I've seen where it is possible to use different slot formats where each slot has some uplink and downlink symbols in it. Could slot formats be set in srsRAN to adjust UL and DL allocation? For example, table 11.1.1-1 in TS 38.213 shows slot format 36 with 3 DL symbols and 10 UL symbols.

**Reply by frankist** [reply]
>Would it be possible to use the TDD-UL-DL-ConfigCommon to set a static configuration? This is discussed in section 11 of TS 38.213. I think it uses the RRC instead of PDCCH DCI. Can srsRAN already do this?

Yes, we only use the static configuration via `TDD-UL-DL-ConfigCommon`. When you are changing the TDD pattern in the config file, you are effectively altering this field in the RRC message. You can confirm this change via the mac or f1ap pcaps generated by the srsgnb.

>Also, I've seen where it is possible to use different slot formats where each slot has some uplink and downlink symbols in it. Could slot formats be set in srsRAN to adjust UL and DL allocation? For example, table 11.1.1-1 in TS 38.213 shows slot format 36 with 3 DL symbols and 10 UL symbols.

Unfortunately, we do not fully support this at the moment. We support PDCCH, PDSCH, and PUCCH in special slots (i.e. slots with some DL and some UL symbols), but not PUSCH. To support it, we would need to support changing the DMRS positioning. Not sure how hard it is to do it.

If you do not mind me asking, is there anything in specific that you are trying to do that requires more esoteric TDD patterns?

**Reply by jasonfarmer0992** [reply]
Thank you for the information @frankist.

I want to upload multiple high resolution video data streams in real time to a server.

If I set the PDSCH to contain some UL symbols, would that increase my UL throughput?

**Reply by herlesupreeth** [reply]
@jasonfarmer0992 

> Also, I've seen where it is possible to use different slot formats where each slot has some uplink and downlink symbols in it. Could slot formats be set in srsRAN to adjust UL and DL allocation? For example, table 11.1.1-1 in TS 38.213 shows slot format 36 with 3 DL symbols and 10 UL symbols.

In order to specify those slot formats DCI 2_0 is needed, which is not yet supported. Also, as @frankist mentioned, currently we support scheduling of PDSCH in slots where some of the symbols are DL but we dont support scheduling of PUSCH in slots where not all symbols are UL.

**Reply by herlesupreeth** [reply, marked as answer]
For now, I would suggest to change the TDD configuration as follows in your gnb.yaml file and give it a try

```
  tdd_ul_dl_cfg:
    dl_ul_tx_period: 10
    nof_dl_slots: 4
    nof_dl_symbols: 9
    nof_ul_slots: 5
    nof_ul_symbols: 0
```

**Reply by jasonfarmer0992** [reply]
@herlesupreeth thank you so much! That increased my uplink speed by about 20 mbps.

**Reply by HaoxinSEU** [reply]
Hi @jasonfarmer0992, 

May I ask one question about your results under this configuration? Now the numbers of slots and symbols allocated to DL and UL are quite close, so I assume their throughput should also be similar. 

But what I got is that DL throughput is higher than UL. Did you get the same results? Or in your tests these two throughput are the same?

Thanks!

**Reply by herlesupreeth** [reply]
btw, you can also use 256QAM MCS table for PUSCH and may get even higher uplink throughput (you could also do the same for PDSCH as well)

```
  pusch:
    mcs_table: qam256
```

**Reply by jasonfarmer0992** [reply]
@HaoxinSEU Yes, my DL throughput is still a little higher than my UL throughput.

My previous throughput was around UL 50 Mbps, DL 110 Mbps.
My new throughput is around UL 70 Mbps, DL 90 Mbps.

Based on the slot configuration, I would have expected them to be the same. My best guess is either a lower quality receive channel on my setup or my EM9191 UE is using a lower modulation scheme for some reason. It could also be some other reason from the 5G spec that I am completely unaware of.

@herlesupreeth I will try the 256QAM table again. I used it before but my SNR was poor due to low gain antennas, so I switched back to 64QAM. I just upgraded to a sector antenna that is way better, so it might work now.

**Reply by HaoxinSEU** [reply]
Thanks for the reply! I guess it's from the spec. Because I tried to use the same MCS for UL and DL by setting `max_ue_mcs` in the configuration, but there is still a gap in the throughput.

**Reply by fahad-ece** [reply]
Dear @herlesupreeth,

For uplink which modulation technique is used? OFDMA or SC-FDMA for srsRAN 5G? Thank you.

**Comment by jasonfarmer0992** [comment]
Thank you to everyone for your replies! I really appreciate your help.

I tried changing the number of DL and UL slots in the `cell-cfg` section, but I still can't get the uplink data rate to increase. However, the downlink data rate will decrease. I seem to be having the same problem as @HaoxinSEU. With my original setup of 3 UL slots, 6 DL slots, SISO antenna setup, and 50 MHz of bandwidth, I can achieve approximately 110 mbps download and 45 mbps upload. This seems to be in line with other users that I have seen on srsRAN forums for the maximum performance of the USRP B210 with 64QAM modulation. No matter what I change in the config file, I cannot increase the uplink speed. I can only decrease the downlink speed. 

I have attached two zip archives. Each zip archive has a config and log file. I set the log `all_level` to `info` to get everything. Each log was generated by running srsRAN with the associated config file and doing a few iperf3 uplink and downlink speed tests using a Sierra Wireless EM9191 modem as the UE. The speed testing results are as follows: 

- 3 uplink slots, 6 downlink slots: UL 45 mbps, DL 110 mbps
- 6 uplink slots, 3 downlink slots: UL 45 mbps, DL 45 mbps

I have tinkered with many other settings in the config file that I thought might be overriding the slot configuration. You can see in the config file where I commented out various settings I have tried. All the other settings I tried did nothing or caused worse performance.
[srsran_config_and_log_3upslots_6downslots.zip](https://github.com/srsran/srsRAN_Project/files/12910751/srsran_config_and_log_3upslots_6downslots.zip)
[srsran_config_and_log_6upslots_3downslots.zip](https://github.com/srsran/srsRAN_Project/files/12910752/srsran_config_and_log_6upslots_3downslots.zip)

**Comment by jasonfarmer0992** [comment]
I appreciate everyone's help on this. Thanks a lot @herlesupreeth and @frankist! The current UL rate can handle my setup now. I will probably have to upgrade my SDR to get more bandwidth for UL and DL in the future. I am going to go ahead and close this for now. If you ever add more slot/symbol configuration options and need help testing, let me know.

**Comment by fllay** [comment]
Hi 
how does this configuration look like?
```
  tdd_ul_dl_cfg:
    dl_ul_tx_period: 10
    nof_dl_slots: 4
    nof_dl_symbols: 9
    nof_ul_slots: 5
    nof_ul_symbols: 0
```

something  like 

DSUUUDSUUD ?

and what `nof_dl_symbols: 9` mean?  Does it mean 9  out of 14 symbols in sub-frame are downlink symbols? if so, which sub-frame this 9 downlink symbols are in?

**Reply by herlesupreeth** [reply]
No, the pattern is as follows DDDDSUUUUU . And those DL symbols are for the S slot, so S slot first 9 symbols are DL and remaining are not assigned

**Reply by fllay** [reply]
Thank you very much. May I ask why do you choose `nof_dl_symbols: 9` for this use case?

**Reply by herlesupreeth** [reply]
You can choose any number of symbols in the range 5-10. The reason for max of 10 DL symbols is to allow sufficient time for switching to UL

**Reply by herlesupreeth** [reply]
Of course the config doesn’t restrict you to 10 DL symbols but it is advised to allow 2-4 symbols for switching from DL to UL

**Reply by fllay** [reply]
thank you. It seems that we can control switching time using `S` slot.
